feat: add configurable additional labels to NodePool metrics

moko-poi opened the pull request:
## What problem does this PR solve?

Closes #2221

Allows users to customize which NodePool labels are included in NodePool metrics (`karpenter_nodepools_limit` and `karpenter_nodepools_usage`), enabling better observability for custom attributes like capacity type, availability zone, and architecture.

## How does this PR solve the problem?

### Implementation approach

Added a CLI flag `--additional-nodepool-metric-labels` (environment variable: `ADDITIONAL_NODEPOOL_METRIC_LABELS`) that accepts a comma-separated list of label keys to include in NodePool metrics.

**Default behavior**: No additional labels (empty by default). Users must explicitly configure which labels to include.

### Key design decisions

1. **CLI flag instead of annotations**: Satisfies Prometheus constraint that all time series for a metric must have identical label keys. Label keys are fixed at startup when metrics are registered.

2. **Deferred initialization**: Follows the same pattern as Node metrics controller - metrics are declared as package variables but initialized in `initializeMetrics()` called from `NewController()`. This ensures options are available when building the label key set.

3. **Empty string for missing labels**: If a NodePool doesn't have a specified label, the metric uses an empty string `""` for that label value. This maintains consistent label keys across all NodePools while allowing flexibility.

4. **No defaults**: By not providing default labels, this PR stays focused on providing the capability without making assumptions about which labels exist on users' NodePools.

### Example usage

```bash
# No additional labels (default)
karpenter

# Add custom labels for capacity type, zone, and architecture
karpenter --additional-nodepool-metric-labels=capacity_type,zone,architecture

# Or use environment variable
export ADDITIONAL_NODEPOOL_METRIC_LABELS=environment,team,version
karpenter
```

With a NodePool:
```
apiVersion: karpenter.sh/v1
kind: NodePool
metadata:
  name: production-spot
  labels:
    capacity_type: spot
    zone: us-east-1a
    architecture: amd64
spec:
  limits:
    cpu: "1000"
```

When configured with --additional-nodepool-metric-labels=capacity_type,zone,architecture, produces metrics:
```
karpenter_nodepools_limit{nodepool="production-spot",resource_type="cpu",capacity_type="spot",zone="us-east-1a",architecture="amd64"} 1000
```

Without the flag, produces metrics:
```
karpenter_nodepools_limit{nodepool="production-spot",resource_type="cpu"} 1000
```

## Testing

- Added comprehensive test coverage in suite_test.go:
  - Verifies additional labels are included in metrics when configured
  - Validates empty string behavior for missing labels
  - All 6 tests passing

## Checklist

- Added CLI flag with empty default
- Implemented deferred metric initialization
- Updated controller wiring to pass context
- Added test helper support
- Comprehensive test coverage (6/6 passing)
- Follows existing Node metrics pattern for consistency

moko-poi committed:
feat: add configurable additional labels to NodePool metrics

grosser commented on pkg/controllers/metrics/nodepool/controller.go:
```diff
@@ -131,11 +146,23 @@ func getLimits(nodePool *v1.NodePool) corev1.ResourceList {
+			labels[labelKey] = value
+		} else {
+			labels[labelKey] = ""
+		}
```
already produces empty string on missing
```suggestion
		labels[labelKey] = nodePool.Labels[labelKey]
```

grosser commented on pkg/operator/options/options.go:
```diff
@@ -157,6 +161,16 @@ func (o *Options) Parse(fs *FlagSet, args ...string) error {
+		})
+	} else {
+		o.AdditionalNodePoolMetricLabels = []string{}
+	}
```
nit: less code paths 🤷 
```suggestion
	o.AdditionalNodePoolMetricLabels = lo.Map(strings.Split(o.rawAdditionalNodePoolMetricLabels, ","), func(s string, _ int) string {
		return strings.TrimSpace(s)
	})
```

grosser reviewed (commented):
logic makes sense
idk about the defaults

moko-poi committed:
refactor: simplify map access in label assignment

moko-poi committed:
refactor: use FilterMap for additional nodepool metric labels parsing

Replace conditional logic with lo.FilterMap to reduce code paths and
filter empty strings. This prevents invalid empty label keys from being
added to metrics when the input contains empty strings or consecutive
commas.

moko-poi committed:
refactor: remove default additional nodepool metric labels

grosser reviewed (approved):
✨

moko-poi commented:
@engedaam @tallaxes 

Could you please review this PR when you have a chance? This feature adds configurable additional labels to NodePool metrics, which has been approved by @grosser. Looking forward to your feedback.

Thank you!
